flask/launcher/src/main/java/net/corda/flask/launcher/JavaProcessBuilder.java

```diff
@@ -0,0 +1,179 @@
+                        break;
+                    default:
+                        sb.append(c);
+                        break;
+                }
+            }
+            sb.append('"');
+            if(++i < strings.size()) {
+                sb.append(' ');
+            }
+        }
+        return sb.toString();
+    }
+
+    @SneakyThrows
+    public int exec() {
+        Process[] process = new Process[1];
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            if(process[0] != null) {
+                if (process[0].isAlive()) {
+                    process[0].destroy();
+                }
+                if (process[0].isAlive()) {
+                    process[0].destroyForcibly();
+                }
+            }
+        }));
+        process[0] = build().inheritIO().start();
+        return process[0].waitFor();
+    }
+
+    @SneakyThrows
+    public ProcessBuilder build() {
+        ArrayList<String> cmd = new ArrayList<>();
+        Path javaBin = Paths.get(javaHome, "bin", "java");
+        cmd.add(javaBin.toString());
+        cmd.addAll(jvmArgs);
+        if(!classpath.isEmpty()) {
+            cmd.add("-cp");
+            cmd.add(classpath.stream().collect(Collectors.joining(PATH_SEPARATOR)));
+        }
+        for(Map.Entry<Object, Object> entry : properties.entrySet()) {
+            cmd.add(String.format("-D%s=%s", entry.getKey(), entry.getValue()));
+        }
+        for(JavaAgent javaAgent : javaAgents) {
+            StringBuilder sb = new StringBuilder("-javaagent:" + javaAgent.jar.toString());
+            String agentArguments = javaAgent.args;
+            if(agentArguments != null) {
+                sb.append('=');
+                sb.append(agentArguments);
+            }
+            cmd.add(sb.toString());
+        }
+        if(executableJar != null) {
+            cmd.add("-jar");
+            cmd.add(executableJar.toString());
+        } else if(mainClassName != null) {
+            cmd.add(mainClassName);
+        } else {
+            throw new IllegalArgumentException(
+                    "Either a main class or the path to an executable jar file have to be specified");
+        }
+        cmd.addAll(cliArgs);
+
+        int cmdLength = 0;
+        for(String part : cmd) {
+            cmdLength += part.length();
+        }
+        //Add space between arguments
+        cmdLength += cmd.size() - 1;
+
+        if(log.isDebugEnabled()) {
+            log.debug("Spawning new process with command line: [{}]",
+                    cmd.stream().map(s -> "\"" + s + "\"").collect(Collectors.joining(", ")));
+        }
+        if(cmdLength < COMMAND_LINE_MAX_SIZE || cmd.size() == 1) {
+            return new ProcessBuilder(cmd);
+        } else {
+            Path argumentFile = Files.createTempFile(PROCESS_BUILDER_PREFIX, ".arg");
+            Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+                @Override
+                @SneakyThrows
```

Comment:
A shutdown hook is not supposed to throw any exceptions. Your `@SneakyThrows` may prevent any subsequent hooks from running, which would be a bug. Please catch and display this exception instead.

Reply:
shutdown hooks run in parallel, not sequentially, each one in a in different thread (note that `addShutdownHook` accepts a `Thread` as a parameter), so if one throws an uncaught exception that doesn't mean the others won't be invoked

you can see this with
```
#!/usr/bin/env groovy
Runtime runtime = Runtime.getRuntime()

runtime.addShutdownHook(new Thread({
  throw new RuntimeException('KA-BOOM!!')
}))

runtime.addShutdownHook(new Thread({
  println('here')
}))
```
the string "here" is always printed...

Reply:
Fair enough. We are just disagreeing on what "best practices" Java code look like then.